src/node_task_runner.cc

```diff
@@ -0,0 +1,219 @@
+#include "node_task_runner.h"
+#include "util.h"
+
+#include <regex>  // NOLINT(build/c++11)
+
+namespace node::task_runner {
+
+#if defined(_WIN32)
+static const char separator[] = "\\";
+static const char bin_path[] = "node_modules\\.bin";
+static const char task_file[] = "cmd.exe";
+#else
+static const char separator[] = "/";
+static const char bin_path[] = "node_modules/.bin";
+static const char task_file[] = "/bin/sh";
+#endif  // defined(_WIN32)
+
+ProcessRunner::ProcessRunner(
+    std::unique_ptr<InitializationResultImpl>* result,
+    std::string_view command,
+    const std::optional<std::string>& positional_args) {
+  process_ = {};
+  loop = uv_default_loop();
+  memset(&options, 0, sizeof(uv_process_options_t));
+
+  // Inherit stdin, stdout, and stderr from the parent process.
+  options.stdio_count = 3;
+  child_stdio[0].flags = UV_INHERIT_FD;
+  child_stdio[0].data.fd = 0;
+  child_stdio[1].flags = UV_INHERIT_FD;
+  child_stdio[1].data.fd = 1;
+  child_stdio[2].flags = UV_INHERIT_FD;
+  child_stdio[2].data.fd = 2;
+  options.stdio = child_stdio;
+  options.exit_cb = ExitCallback;
+  options.file = task_file;
+
+  init_result = result;
+  process_.data = this;
+
+  // Get the current working directory.
+  char cwd[PATH_MAX_BYTES];
+  size_t cwd_size = PATH_MAX_BYTES;
+  CHECK_EQ(uv_cwd(cwd, &cwd_size), 0);
+  CHECK_GT(cwd_size, 0);
+
+  // Example: "/Users/user/project/node_modules/.bin"
+  std::string current_bin_path = cwd + std::string(separator) + bin_path;
+
+  std::string command_str = std::string(command);
+  // Conditionally append the positional arguments to the command string
+  if (positional_args.has_value()) {
+    // Properly escape the positional arguments
+    command_str += " " + EscapeShell(positional_args.value());
+  }
+
+  // Example: "/bin/sh -c 'node test.js'"
+  std::vector<std::string> command_args = {options.file, "-c", command_str};
+  auto argc = command_args.size();
+  CHECK_GE(argc, 1);
+  options.args = new char*[argc + 1];
+  for (size_t i = 0; i < argc; ++i) {
+    options.args[i] = new char[command_args[i].size() + 1];
+    strcpy(options.args[i], command_args[i].c_str());
+  }
+  options.args[argc] = nullptr;
+
+  // Set environment variables
+  uv_env_item_t* env_items;
+  int env_count;
+  CHECK_EQ(uv_os_environ(&env_items, &env_count), 0);
+  options.env = reinterpret_cast<char**>(malloc(sizeof(char*) * (env_count + 1)));
+
+  for (int i = 0; i < env_count; i++) {
+    std::string name = env_items[i].name;
+    std::string value = env_items[i].value;
+    // Check if env_items[i] is matching case-insensitive "path" using regexp
+    if (strcasecmp(env_items[i].name, "path") == 0) {
+      value.insert(0, current_bin_path + ":");
+    }
+    value.insert(0, name + "=");
+    // Environment variables should be in "KEY=value" format
+    options.env[i] = strdup(value.c_str());
```

Comment:
Is this `free()`ed anywhere?

You could also just store these in ProcessRunner so that they just go away together with the ProcessRunner when the uv_run returns

Reply:
Fixed it. Thanks!